src/middleware.mjs

```diff
@@ -147,28 +209,42 @@ export const processRequest = (
                 'Request disconnected before file upload stream parsing.'
               )
             )
-          else if (!upload.done) {
-            upload.file.stream.truncated = true
-            upload.file.stream.emit(
-              'error',
-              new FileStreamDisconnectUploadError(
-                'Request disconnected during file upload stream parsing.'
-              )
-            )
-          }
+
+      if (!parser._finished)
+        parser.destroy(
+          new FileStreamDisconnectUploadError(
+            'Request disconnected during file upload stream parsing.'
+          )
+        )
     })
 
     request.pipe(parser)
   })
 
 export const apolloUploadKoa = options => async (ctx, next) => {
-  if (ctx.request.is('multipart/form-data'))
+  if (!ctx.request.is('multipart/form-data')) return next()
+
+  const finished = new Promise(resolve => ctx.req.on('end', resolve))
+  try {
     ctx.request.body = await processRequest(ctx.req, options)
-  await next()
+    await next()
+  } finally {
+    await finished
+  }
 }
 
 export const apolloUploadExpress = options => (request, response, next) => {
   if (!request.is('multipart/form-data')) return next()
+
+  const finished = new Promise(resolve => request.on('end', resolve))
+  const { send } = response
+  response.send = (...args) => {
```

Comment:
Is this monkey patching the `send` method for all following express middleware? This seems pretty cheeky.

Reply:
Ya, this a really ugly thing to do, but it’s the only way I know to “wrap” child middleware in express/connect without requiring them to be aware of the parent middleware... I’ve reluctantly used this pattern in the past and haven’t really run into issues, but if you know of any other way to do this, I’d love to learn!

Reply:
Of the top of my head it might break `res.write()`: https://stackoverflow.com/questions/44692048/what-is-the-difference-between-res-send-and-res-write-in-express

But I don't know any existing code bases (out of dozens I overview or maintain) which use `res.write()` with express.js.

Reply:
@koresar you're definitely correct that calling `res.write()` could break some assumptions, and also that doing so is a kind of escape hatch from express. In fact, calling `ctx.req.write()` would break koa as well.